File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5ObjectEntry.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.val;
import xyz.wagyourtail.commons.core.CollectionUtils;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Json5ObjectEntry extends StringData.OnlyParsed<Json5ObjectEntry.ObjectEntry> {


    public Json5ObjectEntry(ObjectEntry content) {
        super(content);
    }

    public static Json5ObjectEntry parse(String raw) {
        StringCharReader charReader = new StringCharReader(raw);
        val content = contentBuilder(charReader);
        charReader.expectEOS();
        return new Json5ObjectEntry(content);
    }

    public static Json5ObjectEntry parse(CharReader<?> charReader) {
        return new Json5ObjectEntry(contentBuilder(charReader));
    }

    public static ObjectEntry contentBuilder(CharReader<?> reader) {
        Json5ObjectKey key = new Json5ObjectKey(reader);
        List<Object> postKey = new ArrayList<>();
        while (!reader.exhausted()) {
            val value = reader.parseOrNull(
                    Json5Whitespace::new,
                    Json5Comment::new
            );
            if (value == null) break;
            postKey.add(value);
        }
        reader.expect(':');
        List<Object> preValue = new ArrayList<>();
        while (!reader.exhausted()) {
            val value = reader.parseOrNull(
                    Json5Whitespace::new,
                    Json5Comment::new
            );
            if (value == null) break;
            preValue.add(value);
        }
        Json5Value value = Json5Value.parse(reader);
        return new ObjectEntry(key, postKey, preValue, value);
    }

    public String getKey() {
        return getContent().getKey().getValue();
    }

    public Json5Value getValue() {
        return getContent().getValue();
    }

    @Getter
    @AllArgsConstructor
    public static class ObjectEntry extends Data.Content {
        private final Json5ObjectKey key;
        private final List<Object> postKey;
        private final List<Object> preValue;
        private final Json5Value value;

        @Override
        public Iterable<Object> getEntries() {
            return CollectionUtils.concat(
                    Collections.singleton(key),
                    postKey,
                    Collections.singleton(":"),
                    preValue,
                    Collections.singleton(value)
            );
        }

    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5ObjectEntry.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader
import xyz.wagyourtail.commonskt.utils.iterable

class Json5ObjectEntry(content: ObjectEntry) : StringData.OnlyParsed<Json5ObjectEntry.ObjectEntry>(content) {

    companion object : StringDataBuilder<Json5ObjectEntry, ObjectEntry> {

        override fun invoke(rawContent: CharReader<*>): Json5ObjectEntry {
            return Json5ObjectEntry(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): ObjectEntry {
            val key = Json5ObjectKey(reader)
            val postKey = buildList {
                while(!reader.exhausted()) {
                    val value = reader.parseOrNull(
                        Json5Whitespace::invoke,
                        Json5Comment::invoke
                    )
                    if (value == null) break
                    add(value)
                }
            }
            reader.expect(':')
            val preValue = buildList {
                while(!reader.exhausted()) {
                    val value = reader.parseOrNull(
                        Json5Whitespace::invoke,
                        Json5Comment::invoke
                    )
                    if (value == null) break
                    add(value)
                }
            }
            val value = Json5Value(reader)
            return ObjectEntry(key, postKey, preValue, value)
        }

    }

    val key: String
        get() = content.key.value

    val value: Json5Value
        get() = content.value

    class ObjectEntry(
        val key: Json5ObjectKey,
        val postKey: List<Any>,
        val preValue: List<Any>,
        val value: Json5Value
    ) : Content() {

        override val entries: Iterable<*>
            get() = iterable {
                yield(key)
                yieldAll(postKey)
                yield(":")
                yieldAll(preValue)
                yield(value)
            }

    }

}